presto-spark-base/src/main/java/com/facebook/presto/spark/execution/http/PrestoSparkHttpTaskClient.java

```diff
@@ -169,8 +170,30 @@ public ListenableFuture<BaseResponse<TaskInfo>> getTaskInfo()
                 .setHeader(PRESTO_MAX_WAIT, infoRefreshMaxWait.toString())
                 .setUri(taskUri)
                 .build();
-        ResponseHandler responseHandler = createAdaptingJsonResponseHandler(taskInfoCodec);
-        return httpClient.executeAsync(request, responseHandler);
+
+        return httpClient.executeAsync(request, new ResponseHandler<BaseResponse<TaskInfo>, RuntimeException>()
+        {
+            @Override
+            public BaseResponse<TaskInfo> handleException(Request request, Exception exception)
+            {
+                throw propagate(request, exception);
+            }
+
+            @Override
+            public BaseResponse<TaskInfo> handle(Request request, Response response)
+            {
+                if (familyForStatusCode(response.getStatusCode()) != HttpStatus.Family.SUCCESSFUL) {
+                    throw new RuntimeException(format("Unexpected taskInfo fetch response code: %s", response.getStatusCode()));
+                }
+
+                try {
+                    return createAdaptingJsonResponseHandler(taskInfoCodec).handle(request, response);
+                }
+                catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        });
```

Comment:
This seems repetitive with line 126-143. Can you extract this out into a class and reuse in both places. In-fact, be doing this for pretty much all idempotent http api calls.